### ISSUE TITLE
Use a proper argument parsing library

### ISSUE THREAD

**Issue by ReedOei** [ReedOei/Enki#1]
Currently argument parsing is very ad-hoc. Instead, we should use something like http://hackage.haskell.org/package/optparse-applicative

**Comment by JKTKops** [comment]
Personally I'd recommend getopt but it seems like you're doing something fancy so that the command line reads like English. If it gets too fancy, optparse-applicative might not be able to handle it, and then I would recommend parsing it with MegaParsec.

**Comment by ReedOei** [comment]
Yeah, that was just for fun (the English-like commands), I plan to get rid of that. It doesn't really seem appropriate for a command line tool. As for getopt I've mainly mentioned optparse-applicative because it's what I always use, so curious to hear your thoughts on why use getopt as I've never done anything with it.

**Comment by JKTKops** [comment]
Getopt let's you represent each flag as. Haskell term however you want. In
particular, you can represent each flag as a function on some configuration
type in a very succinct and direct way, collect all the specified flags and
fold over applying the functions.

But I also have much more experience with it than optparse, so I'm curious
too :)

**Comment by ReedOei** [comment]
Looking at the two libraries, I'm not sure there's that much of a difference between them. 

Generally for optparse you represent your configuration as some type, then you use the primitive parsers  (e.g., for strings, numbers, whatever) and combinaotrs to build up a parser for your configuration type.

get-opt seems similar, except it doesn't use Applicative to build up the parsers but you have to fold over it yourself. optparse feels more "Haskell-y" to me, but I'm not sure if that's really a huge benefit.

**Comment by JKTKops** [comment]
Interesting. Getopt feels more "Haskell-y" to me because it builds the parser for you given the data that represents the various options you have.

Representing each option as a function and folding over it such a common use case that I don't know why a function to do it isn't included in the library.

The most obvious trade-off to me is that getopt forces your arguments to be in a "--long-form" or "-l" format, so if you want a different argument syntax (most compilers have single-dash arguments with more than one character, for various reasons) then it won't work.
